keps/sig-node/4816-dra-prioritized-list/README.md

```diff
@@ -0,0 +1,1048 @@
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal adds a new field to the `DeviceRequest`, called `FirstOf` which
+will contain an ordered list of `DeviceRequest` objects. In order to satisfy the
+main (containing) request, exactly one of the requests listed in `FirstOf` must
+be satisfied. The order listed is considered a priority order, such that the
+scheduler will only try to use the second item in the list if it is unable to
+satsify the first item, and so on.
+
+A `DeviceRequest` that populates the `FirstOf` field must *not* populate the
+`DeviceClassName` field. The `required` validation on this field will be
+relaxed. This allows existing clients to differentiate between claims they
+understand (with `DeviceClassName`) and those they do not (without
+`DeviceClassName` but with the new field). Clients written for 1.31, when
+`DeviceClassName` was required, were requested to include this logic, and the
+in-tree components have been built in this way.
+
+```go
+// DeviceRequest is a request for devices required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for several identical devices.
+type DeviceRequest struct {
+    // Name can be used to reference this request in a pod.spec.containers[].resources.claims
+    // entry and in a constraint of the claim.
+    //
+    // Must be a DNS label.
+    //
+    // +required
+    Name string
+
+    // DeviceClassName references a specific DeviceClass, which can define
+    // additional configuration and selectors to be inherited by this
+    // request.
+    //
+    // Either a class or FirstOf requests are required in DeviceClaim.Requests.
+    // When this request is part of the FirstOf list, a class is required. Nested
+    // FirstOf requests are not allowed
+    //
+    // Which classes are available depends on the cluster.
+    //
+    // Administrators may use this to restrict which devices may get
+    // requested by only installing classes with selectors for permitted
+    // devices. If users are free to request anything without restrictions,
+    // then administrators can create an empty DeviceClass for users
+    // to reference.
+    //
+    // +optional
+    // +oneOf=deviceRequestType
+    DeviceClassName string
+
+    // FirstOf contains subrequests, exactly one of which must be satisfied
+    // in order to satisfy this request. This field may only be set in the
+    // entries of DeviceClaim.Requests. It must not be set in DeviceRequest
+    // instances that themselves are part of a FirstOf.
+    //
+    // +optional
+    // +oneOf=deviceRequestType
+    FirstOf []DeviceRequest
```

Comment:
Doesn't this call for a dedicated type that doesn't have this field?

Reply:
We're enforcing that in validation. We could do it as a separate type, if that's the consensus.

Reply:
It can be achieved with composition using the json inline tag, though. But that's an implementation detail

Reply:
Can the actual Go type details be sorted out at implementation time? I'd love to use a composition with json inlining, but we'll need to verify that that doesn't cause issues.